vm/src/obj/objdict.rs

```diff
@@ -44,13 +45,28 @@ impl PyDictRef {
         kwargs: KwArgs,
         vm: &VirtualMachine,
     ) -> PyResult<PyDictRef> {
-        let mut dict = DictContentType::default();
+        let dict = DictContentType::default();
 
+        let entries = RefCell::new(dict);
+        // it's unfortunate that we can't abstract over RefCall, as we should be able to use dict
+        // directly here, but that would require generic associated types
+        PyDictRef::merge(&entries, dict_obj, kwargs, vm)?;
+
+        PyDict { entries }.into_ref_with_type(vm, class)
+    }
+
+    fn merge(
+        dict: &RefCell<DictContentType>,
+        dict_obj: OptionalArg<PyObjectRef>,
+        kwargs: KwArgs,
+        vm: &VirtualMachine,
+    ) -> PyResult<()> {
         if let OptionalArg::Present(dict_obj) = dict_obj {
             let dicted: PyResult<PyDictRef> = dict_obj.clone().downcast();
             if let Ok(dict_obj) = dicted {
-                for (key, value) in dict_obj.get_key_value_pairs() {
-                    dict.insert(vm, &key, value)?;
+                let mut dict_borrowed = dict.borrow_mut();
```

Comment:
If this borrow moved inside the loop then would we avoid the borrow error and special case in update below?

Reply:
Not exactly, the dict_obj is still borrowed throughout the loop as the `Ref` isn't dropped and for `iter_items` it can't be dropped before the loop, as it borrows new items on each iteration for the dict. If I reverted back to using `get_key_value_pairs`, that special case would be avoided though.

Reply:
Oh yes, of course, the other borrow.

Once I get my dict iterator patch merged we might be able to use that to iterate the dictionary without keeping the borrow - but it's probably not worth it just to avoid the special case.